File: src/main/java/com/example/reservation/domain/room/Property.java
package com.example.reservation.domain.room;

import com.example.reservation.domain.guest.Address;
import jakarta.persistence.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.*;

@Entity
@Table(name = "properties")
@EntityListeners(AuditingEntityListener.class)
public class Property {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(nullable = false, length = 200)
    private String name;
    
    @Column(length = 1000)
    private String description;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PropertyType type;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PropertyCategory category;
    
    @Column(nullable = false)
    private Integer starRating;
    
    @Embedded
    private Address address;
    
    @Column(length = 20)
    private String phoneNumber;
    
    @Column(length = 320)
    private String email;
    
    @Column(length = 500)
    private String website;
    
    @Embedded
    private CheckInOutInfo checkInOutInfo = new CheckInOutInfo();
    
    @Embedded
    private PropertyPolicies policies = new PropertyPolicies();
    
    @ElementCollection
    @Enumerated(EnumType.STRING)
    @CollectionTable(name = "property_amenities", joinColumns = @JoinColumn(name = "property_id"))
    @Column(name = "amenity")
    private Set<PropertyAmenity> amenities = new HashSet<>();
    
    @OneToMany(mappedBy = "property", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Room> rooms = new ArrayList<>();
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PropertyStatus status = PropertyStatus.ACTIVE;
    
    @Column(nullable = false)
    private Boolean isPublished = false;
    
    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @LastModifiedDate
    @Column(nullable = false)
    private LocalDateTime lastModifiedAt;
    
    protected Property() {}
    
    public Property(String name, PropertyType type, PropertyCategory category, 
                   Integer starRating, Address address) {
        this.name = name;
        this.type = type;
        this.category = category;
        this.starRating = starRating;
        this.address = address;
        this.createdAt = LocalDateTime.now();
        this.lastModifiedAt = LocalDateTime.now();
    }
    
    // 비즈니스 메서드
    public int getTotalRooms() {
        return rooms.size();
    }
    
    public List<Room> getAvailableRooms() {
        return rooms.stream()
                   .filter(Room::isAvailable)
                   .toList();
    }
    
    public List<Room> getRoomsByType(RoomType roomType) {
        return rooms.stream()
                   .filter(room -> room.getType() == roomType)
                   .toList();
    }
    
    public Optional<BigDecimal> getLowestRate() {
        return rooms.stream()
                   .map(Room::getBaseRate)
                   .min(BigDecimal::compareTo);
    }
    
    public Optional<BigDecimal> getHighestRate() {
        return rooms.stream()
                   .map(Room::getBaseRate)
                   .max(BigDecimal::compareTo);
    }
    
    public boolean isAcceptingReservations() {
        return status == PropertyStatus.ACTIVE && isPublished;
    }
    
    public void addRoom(Room room) {
        rooms.add(room);
    }
    
    public void removeRoom(Room room) {
        rooms.remove(room);
    }
    
    public void addAmenity(PropertyAmenity amenity) {
        amenities.add(amenity);
    }
    
    public void removeAmenity(PropertyAmenity amenity) {
        amenities.remove(amenity);
    }
    
    public boolean hasAmenity(PropertyAmenity amenity) {
        return amenities.contains(amenity);
    }
    
    public void publish() {
        this.isPublished = true;
    }
    
    public void unpublish() {
        this.isPublished = false;
    }
    
    public void activate() {
        this.status = PropertyStatus.ACTIVE;
    }
    
    public void deactivate() {
        this.status = PropertyStatus.INACTIVE;
    }
    
    // Getters and Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public PropertyType getType() { return type; }
    public void setType(PropertyType type) { this.type = type; }
    
    public PropertyCategory getCategory() { return category; }
    public void setCategory(PropertyCategory category) { this.category = category; }
    
    public Integer getStarRating() { return starRating; }
    public void setStarRating(Integer starRating) { this.starRating = starRating; }
    
    public Address getAddress() { return address; }
    public void setAddress(Address address) { this.address = address; }
    
    public String getPhoneNumber() { return phoneNumber; }
    public void setPhoneNumber(String phoneNumber) { this.phoneNumber = phoneNumber; }
    
    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }
    
    public String getWebsite() { return website; }
    public void setWebsite(String website) { this.website = website; }
    
    public CheckInOutInfo getCheckInOutInfo() { return checkInOutInfo; }
    public void setCheckInOutInfo(CheckInOutInfo checkInOutInfo) { this.checkInOutInfo = checkInOutInfo; }
    
    public PropertyPolicies getPolicies() { return policies; }
    public void setPolicies(PropertyPolicies policies) { this.policies = policies; }
    
    public Set<PropertyAmenity> getAmenities() { return new HashSet<>(amenities); }
    public void setAmenities(Set<PropertyAmenity> amenities) { this.amenities = new HashSet<>(amenities); }
    
    public List<Room> getRooms() { return new ArrayList<>(rooms); }
    public void setRooms(List<Room> rooms) { this.rooms = new ArrayList<>(rooms); }
    
    public PropertyStatus getStatus() { return status; }
    public void setStatus(PropertyStatus status) { this.status = status; }
    
    public Boolean getIsPublished() { return isPublished; }
    public void setIsPublished(Boolean isPublished) { this.isPublished = isPublished; }
    
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    
    public LocalDateTime getLastModifiedAt() { return lastModifiedAt; }
    public void setLastModifiedAt(LocalDateTime lastModifiedAt) { this.lastModifiedAt = lastModifiedAt; }
    
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Property property = (Property) obj;
        return Objects.equals(id, property.id) && id != null;
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
    
    @Override
    public String toString() {
        return "Property{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", type=" + type +
                ", starRating=" + starRating +
                ", status=" + status +
                '}';
    }
}

File: src/main/kotlin/com/example/reservation/domain/room/Property.kt
package com.example.reservation.domain.room

import com.example.reservation.domain.guest.Address
import jakarta.persistence.*
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.math.BigDecimal
import java.time.LocalDateTime
import java.time.LocalTime

@Entity
@Table(name = "properties")
@EntityListeners(AuditingEntityListener::class)
data class Property(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,
    
    @Column(nullable = false, length = 200)
    val name: String,
    
    @Column(length = 1000)
    val description: String? = null,
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val type: PropertyType,
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val category: PropertyCategory,
    
    @Column(nullable = false)
    val starRating: Int,
    
    @Embedded
    val address: Address,
    
    @Column(length = 20)
    val phoneNumber: String? = null,
    
    @Column(length = 320)
    val email: String? = null,
    
    @Column(length = 500)
    val website: String? = null,
    
    @Embedded
    val checkInOutInfo: CheckInOutInfo = CheckInOutInfo(),
    
    @Embedded
    val policies: PropertyPolicies = PropertyPolicies(),
    
    @ElementCollection
    @Enumerated(EnumType.STRING)
    @CollectionTable(name = "property_amenities", joinColumns = [JoinColumn(name = "property_id")])
    @Column(name = "amenity")
    val amenities: Set<PropertyAmenity> = emptySet(),
    
    @OneToMany(mappedBy = "property", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    val rooms: List<Room> = emptyList(),
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val status: PropertyStatus = PropertyStatus.ACTIVE,
    
    @Column(nullable = false)
    val isPublished: Boolean = false,
    
    @CreatedDate
    @Column(nullable = false, updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),
    
    @LastModifiedDate
    @Column(nullable = false)
    val lastModifiedAt: LocalDateTime = LocalDateTime.now()
) {
    fun getTotalRooms(): Int = rooms.size
    
    fun getAvailableRooms(): List<Room> = rooms.filter { it.isAvailable() }
    
    fun getRoomsByType(roomType: RoomType): List<Room> = rooms.filter { it.type == roomType }
    
    fun getLowestRate(): BigDecimal? = rooms.minOfOrNull { it.baseRate }
    
    fun getHighestRate(): BigDecimal? = rooms.maxOfOrNull { it.baseRate }
    
    fun isAcceptingReservations(): Boolean = 
        status == PropertyStatus.ACTIVE && isPublished
        
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Property) return false
        return id != 0L && id == other.id
    }
    
    override fun hashCode(): Int = id.hashCode()
}

@Embeddable
data class CheckInOutInfo(
    @Column(nullable = false)
    val checkInTime: LocalTime = LocalTime.of(15, 0),
    
    @Column(nullable = false)  
    val checkOutTime: LocalTime = LocalTime.of(11, 0),
    
    @Column(nullable = false)
    val lateCheckInAllowed: Boolean = true,
    
    @Column(nullable = false)
    val earlyCheckInAllowed: Boolean = true,
    
    @Column(nullable = false)
    val lateCheckOutAllowed: Boolean = true
)

@Embeddable
data class PropertyPolicies(
    @Column(nullable = false)
    val cancellationPolicy: String = "24시간 전 무료 취소",
    
    @Column(nullable = false)
    val petPolicy: String = "반려동물 동반 불가",
    
    @Column(nullable = false)
    val smokingPolicy: String = "전 객실 금연",
    
    @Column(nullable = false)
    val childPolicy: String = "만 18세 미만 투숙 불가",
    
    @Column(nullable = false)
    val extraBedPolicy: String = "추가 침대 요청 가능",
    
    @Column(nullable = false)
    val ageRestriction: Int = 18
)

enum class PropertyType {
    HOTEL, RESORT, MOTEL, PENSION, GUESTHOUSE, HOSTEL, APARTMENT, VILLA, CAMPING
}

enum class PropertyCategory {
    LUXURY, PREMIUM, STANDARD, BUDGET, BUSINESS, BOUTIQUE, FAMILY, ROMANTIC
}

enum class PropertyAmenity {
    WIFI, PARKING, POOL, GYM, SPA, RESTAURANT, BAR, ROOM_SERVICE, 
    CONCIERGE, BUSINESS_CENTER, MEETING_ROOM, AIRPORT_SHUTTLE,
    LAUNDRY, DRY_CLEANING, SAFE_DEPOSIT_BOX, CURRENCY_EXCHANGE,
    ELEVATOR, WHEELCHAIR_ACCESS, PET_FRIENDLY, NON_SMOKING,
    AIR_CONDITIONING, HEATING, BALCONY, GARDEN, BEACH_ACCESS
}

enum class PropertyStatus {
    ACTIVE, INACTIVE, MAINTENANCE, CLOSED_TEMPORARILY, CLOSED_PERMANENTLY
}